### PR TITLE
feat(aws-fargate-s3): Created new construct

### LINKED ISSUE #590: feat(aws-fargate-s3): New Construct
Create a construct that links a Fargate Service to an S3 Bucket

Should create a VPC or accept an existing VPC
Should create a Fargate Service or accept an existing service
Should create an S3 bucket or accept an existing bucket
Should grant the service IAM permission to send and delete messages on the queue
Should set up environment variables in the service that identify the S3 Bucket in the construct code so the container can read/write to the correct bucket

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by mickychetta**
*Issue #590 , if available:*

*Description of changes:*
-Created new construct of aws-fargate-s3
-Added environment variables in fargate environment to allow read/write to bucket
-Created unit/integ tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

fixes #590

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 6265f151ea455ef1d9864719654809ce7696d0b9
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=82f20f44-e73a-49cb-9a63-af0eb7db77ec%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 6265f151ea455ef1d9864719654809ce7696d0b9
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6f122101-c9ee-450d-b111-71a51b1d8978%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: aec48ca35173bb27dc5ef837ff05fbc9b326db92
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=31c7dd89-0785-4c55-ab88-612e237ab9f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: aec48ca35173bb27dc5ef837ff05fbc9b326db92
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8c2a8ea3-28fa-4acb-81c1-5017f4020087%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 845bd7c5e2b6d46781c12b51bd2aa58126b4eca8
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=398b3b51-a266-45cf-a6d1-cadb34f1efe8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 845bd7c5e2b6d46781c12b51bd2aa58126b4eca8
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1e7754c6-e89d-437f-9f9e-ef6f97138bb5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: dd8c5ea747ba62a4d9158740b6b1b808878c6a21
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=38d9f0a7-80d8-41bc-bf4a-23b20ca43dc4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: dd8c5ea747ba62a4d9158740b6b1b808878c6a21
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=10350979-0a74-43d0-89c5-a18676ad75af%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 19a7f594aa9cb0b233829c1e036f41c74f70b192
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8f232368-b4d7-456e-a3d3-c70d8ec97cc1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 19a7f594aa9cb0b233829c1e036f41c74f70b192
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d4dbee90-a15a-468a-a99b-d345671d6b5b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 07e432281816c3a9a6b361e263ab9911a22d8003
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8c670600-9b60-4426-80d9-af1b5af08961%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 07e432281816c3a9a6b361e263ab9911a22d8003
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=21fc5571-c3f6-4be4-93cf-c68839dce938%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-fargate-s3/test/fargate-s3.test.ts [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+    publicApi,
+    ecrRepositoryArn: defaults.fakeEcrRepoArn,
+    vpcProps: { cidr: '172.0.0.0/16' },
+    clusterProps: { clusterName },
+    containerDefinitionProps: { containerName },
+    fargateTaskDefinitionProps: { family: familyName},
+    fargateServiceProps: { serviceName },
+    bucketProps: { bucketName },
+    logS3AccessLogs: false,
+    bucketPermissions: ['Delete', 'Put', 'Read', 'ReadWrite', 'Write']
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-s3/test/fargate-s3.test.ts:41]: There's no check on these in any of the tests.  Including all the options will run al the code and get you 100% coverage, but without confirmation of the output it's just a gimmick to increase the coverage total.

One test should specify a subset, then confirm the result (probably have to confirm with a non-granular check like line 75.

**Review by biffgaut — CHANGES_REQUESTED** [review]
Similar issue around permissions. Change should very simple.

Testing for conditions needs improvement.

File: source/patterns/@aws-solutions-constructs/aws-fargate-s3/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+|existingBucketInterface?|[`s3.IBucket`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.IBucket.html)|Existing S3 Bucket interface. Providing this property and `bucketProps` results in an error.|
+|bucketProps?|[`s3.BucketProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.BucketProps.html)|Optional user provided props to override the default props for the S3 Bucket.|
+|loggingBucketProps?|[`s3.BucketProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.BucketProps.html)|Optional user provided props to override the default props for the S3 Logging Bucket.|
+|logS3AccessLogs?| boolean|Whether to turn on Access Logging for the S3 bucket. Creates an S3 bucket with associated storage costs for the logs. Enabling Access Logging is a best practice. default - true|
+|bucketPermissions?|`string[]`|Optional bucket permissions to grant to the Lambda function. One or more of the following may be specified: `Delete`, `Put`, `Read`, `ReadWrite`, `Write`.|
```
[the remark this answers is not in the document]
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/aws-fargate-s3/README.md:71]: The default bucket permissions in aws-lambda-s3 is `ReadWrite` which composes of Read, Write, and Delete operations so I will stick to to that for consistency. I allowed Read, Write, and Delete strings for bucketPermissions  in order to create a more fine-grained policy. I removed Put as it is repetitive and made ReadWrite as default which separates confusion from inputs.

[the remark this answers is not in the document]
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/aws-fargate-s3/README.md:71]: We will need to make following changes to aws-lambda-s3 if we agree with this.

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 27b900622b57f4e697368c9b5e52b9543b8ac152
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=07cb4ae0-7277-4d84-b574-ac27f41eca03%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 27b900622b57f4e697368c9b5e52b9543b8ac152
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2725a532-9fb9-4ce2-b907-7c0751af04a1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: f14cc541810f30484a916488fcbc4d9767b2c921
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8977a01c-dfc2-458c-a642-f7d4b12d1b72%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-fargate-s3/README.md [hunk 3]
```diff
@@ -68,7 +56,7 @@ _Parameters_
 |bucketProps?|[`s3.BucketProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.BucketProps.html)|Optional user provided props to override the default props for the S3 Bucket.|
 |loggingBucketProps?|[`s3.BucketProps`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-s3.BucketProps.html)|Optional user provided props to override the default props for the S3 Logging Bucket.|
 |logS3AccessLogs?| boolean|Whether to turn on Access Logging for the S3 bucket. Creates an S3 bucket with associated storage costs for the logs. Enabling Access Logging is a best practice. default - true|
-|bucketPermissions?|`string[]`|Optional bucket permissions to grant to the Lambda function. One or more of the following may be specified: `Delete`, `Put`, `Read`, `ReadWrite`, `Write`.|
+|bucketPermissions?|`string[]`|Optional bucket permissions to grant to the Fargate service. One or more of the following may be specified: `Delete`, `Read`, and `Write`. Default is `ReadWrite` which includes `[s3:GetObject*, s3:GetBucket*, s3:List*, s3:DeleteObject*, s3:PutObject*, s3:Abort*]`.|
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-s3/README.md:59]: Need to change this to Default is 'Read' and 'Write' as we don't want to give the impression that ReadWrite is not deprecated. I like that we don't list it here - I assume the code still allows it for backwards compatibility.

If the code looks good I can approve this now and we can make a quick inline change of this comment later.

File: source/patterns/@aws-solutions-constructs/aws-fargate-s3/lib/index.ts [hunk 4]
```diff
@@ -208,20 +209,24 @@ export class FargateToS3 extends Construct {
       if (props.bucketPermissions.includes('Delete')) {
         bucket.grantDelete(this.service.taskDefinition.taskRole);
       }
-      if (props.bucketPermissions.includes('Put')) {
-        bucket.grantPut(this.service.taskDefinition.taskRole);
-      }
       if (props.bucketPermissions.includes('Read')) {
         bucket.grantRead(this.service.taskDefinition.taskRole);
       }
-      if (props.bucketPermissions.includes('ReadWrite')) {
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-s3/lib/index.ts]: We can't delete this functionality as that will break existing clients - we should drop them from the doc but leave the code that supports them in place along with comments that 'Put' and 'ReadWrite' are deprecated but this code remains for backwards compatibility.

```
   // ReadWrite and Put are deprecated, but included here for backwards compatibility
   if (props.bucketPermissions.includes('Write') || 
        props.bucketPermissions.includes{'ReadWrite') ||
        props.bucketPermissions.includes('Put') )
   {
        ...
   }
   if (props.bucketPermissions.includes('Read') || props.bucketPermissions.include('ReadWrite') {
      ...
   }

etc.

  if (props.bucketPermissions.includes('ReadWrite') || props.bucketPermissions.includes('Put')) {
    printWarning('Put and ReadWrite are deprecated, please switch to some combination of Read, Write and Delete');
  }

```

**Review by biffgaut — CHANGES_REQUESTED** [review]
Couple of changes to ensure backwards compatibility with existing stacks.

**Review by biffgaut — APPROVED** [review]
Ignore my comments about backwards compatibility - I was thinking aws-lambda-s3 and this is a brand new construct.

**Pull request merged**